examples/gopherjs-humble/go/views/app.go

```diff
@@ -0,0 +1,155 @@
+package views
+
+import (
+	"strings"
+
+	"github.com/go-humble/examples/todomvc/go/models"
+	"github.com/go-humble/examples/todomvc/go/templates"
+	"github.com/go-humble/temple/temple"
+	"github.com/go-humble/view"
+	"honnef.co/go/js/dom"
+)
+
+const (
+	// Constants for certain keycodes.
+	enterKey  = 13
+	escapeKey = 27
+)
+
+var (
+	appTmpl  = templates.MustGetTemplate("app")
+	document = dom.GetWindow().Document()
+)
+
+// App is the main view for the application.
+type App struct {
+	Todos *models.TodoList
+	tmpl  *temple.Template
+	// predicate will be used to filter the todos when rendering. Only
+	// todos for which the predicate is true will be rendered.
+	predicate models.Predicate
+	view.DefaultView
+}
+
+// UseFilter causes the app to use the given predicate to filter the todos when
+// rendering. Only todos for which the predicate returns true will be rendered.
+func (v *App) UseFilter(predicate models.Predicate) {
+	v.predicate = predicate
+}
+
+// NewApp creates and returns a new App view, using the given todo list.
+func NewApp(todos *models.TodoList) *App {
+	v := &App{
+		Todos: todos,
+		tmpl:  appTmpl,
+	}
+	v.SetElement(document.QuerySelector(".todoapp"))
+	return v
+}
+
+// tmplData returns the data that is passed through to the template for the
+// view.
+func (v *App) tmplData() map[string]interface{} {
+	return map[string]interface{}{
+		"Todos": v.Todos,
+		"Path":  dom.GetWindow().Location().Hash,
+	}
+}
+
+// Render renders the App view and satisfies the view.View interface.
+func (v *App) Render() error {
+	if err := v.tmpl.ExecuteEl(v.Element(), v.tmplData()); err != nil {
+		return err
+	}
+	listEl := v.Element().QuerySelector(".todo-list")
+	for _, todo := range v.Todos.Filter(v.predicate) {
+		todoView := NewTodo(todo)
+		// NOTE: the todomvc tests require that the top-level element for the
+		// todo view is an li element. Unfortunately there is no way to express
+		// this while also having template logic determine whether or not the
+		// li element should have the class "completed". I would prefer to have
+		// the top-level element for each todo be a div wrapper, and to include
+		// the li element inside the template. The workaround for now is to create
+		// the li element and set it's class manually.
+		todoView.SetElement(document.CreateElement("li"))
```

Comment:
Oh, I think the above answers my earlier question about the commented out `<li>` in the template. Perhaps you should add a note there that points to this note to explain it.